plugins/experimental/server_push_preload/server_push_preload.cc

```diff
@@ -0,0 +1,109 @@
+#include <iostream>
+#include <regex>
+#include <sstream>
+#include <vector>
+#include <ts/ts.h>
+#include <ts/experimental.h>
+#include <atscppapi/GlobalPlugin.h>
+#include <atscppapi/utils.h>
+
+#define PLUGIN_NAME "server_push_preload"
+
+using namespace std;
+using namespace atscppapi;
+
+static regex linkRegexp("<([^>]+)>;.*rel=preload");
+
+namespace
+{
+GlobalPlugin *plugin;
+}
+
+class LinkServerPushPlugin : public GlobalPlugin
+{
+public:
+  LinkServerPushPlugin()
+  {
+    TSDebug(PLUGIN_NAME, "registering transaction hooks");
+    LinkServerPushPlugin::registerHook(HOOK_SEND_RESPONSE_HEADERS);
+  }
+
+  virtual void
+  handleSendResponseHeaders(Transaction &transaction)
+  {
+    serverPush(transaction);
+    transaction.resume();
+  }
+
+  void
+  serverPush(Transaction &transaction)
+  {
+    ClientRequest &request = transaction.getClientRequest();
+    if (request.getMethod() != HTTP_METHOD_GET) {
+      return;
+    }
+
+    Response &response = transaction.getClientResponse();
+    Headers &headers = response.getHeaders();
+
+    TSHttpTxn txnp = static_cast<TSHttpTxn>(transaction.getAtsHandle());
+    string linkHeader = headers.values("Link");
+    const Url &clientUrl = request.getPristineUrl();
+
+    if (linkHeader.empty()) {
+      TSDebug(PLUGIN_NAME, "Link header not found for url: %s", clientUrl.getUrlString().c_str());
+      return;
+    }
+
+    TSDebug(PLUGIN_NAME, "Found Link header: %s", linkHeader.c_str());
+    stringstream ss(linkHeader);
+    vector<string> links = split(linkHeader, ',');
+
+    for (const string& link : links) {
+      TSDebug(PLUGIN_NAME, "Link: %s", link.c_str());
+      if (link.find("nopush") != string::npos) {
```

Comment:
This check should be a bit more strict. "nopush" might be found in the href attribute. Also, other strings (e.g. "volcanopush", "nopushpin") will pass this check. Probably we should find " nopush" ("no push" with a leading whitespace) and check if the following character is ';' or end of string.